src/main/scala/viper/gobra/frontend/ParseTreeTranslator.scala

```diff
@@ -0,0 +1,2240 @@
+        // =
+        val assignees = visitAssigneeList(ctx.rangeClause().expressionList()) match {
+          case v : Vector[PAssignee] => v
+          case _ => fail(ctx)
+        }
+        PAssForRange(range, assignees, block).at(specCtx)
+      }
+    } else {
+      // for { }
+      PForStmt(None, PBoolLit(true).at(ctx.FOR()), None, spec, block).at(specCtx)
+    }
+
+  }
+
+  /**
+    * {@inheritDoc  }
+    *
+    * <p>The default implementation returns the result of calling
+    * {@link #visitChildren} on {@code ctx}.</p>
+    */
+  override def visitTerminationMeasure(ctx: TerminationMeasureContext): PTerminationMeasure = {
+    val cond = visitNodeOrNone[PExpression](ctx.expression())
+    visitExpressionList(ctx.expressionList()) match {
+      case Vector(PBlankIdentifier()) => PWildcardMeasure(cond).at(ctx)
+      case exprs@Vector(_, _*) => PTupleTerminationMeasure(exprs, cond).at(ctx)
+      case Vector() => PTupleTerminationMeasure(Vector.empty, cond).at(ctx.parent match {
+        case s : SpecStatementContext => s.DEC()
+        case l : LoopSpecContext => l.DEC()
+      })
+    }
+  }
+
+  /** Visits a loop specification.
+    *
+    * @param ctx the loop spec context
+    * @return a positioned PLoopSpec node
+    */
+  override def visitLoopSpec(ctx: LoopSpecContext): PLoopSpec = {
+    val invs = for (inv <- ctx.expression().asScala.toVector) yield visitNode[PExpression](inv)
+    val decs = if (has(ctx.terminationMeasure())) Some(visitTerminationMeasure(ctx.terminationMeasure())).pos() else None
+
+    PLoopSpec(invs, decs).at(ctx)
+  }
+  //endregion
+
+  //region Communicaton statements
+  /**
+    * {@inheritDoc  }
+    *
+    * <p>The default implementation returns the result of calling
+    * {@link #   visitChildren} on {@code ctx}.</p>
+    */
+  override def visitGoStmt(ctx: GoStmtContext): PGoStmt = {
+    visitChildren(ctx) match {
+      case Vector("go", exp : PExpression) => PGoStmt(exp).at(ctx)
+    }
+  }
+
+  //region Select statements
+  /**
+    * {@inheritDoc  }
+    *
+    * <p>The default implementation returns the result of calling
+    * {@link #   visitChildren} on {@code ctx}.</p>
+    */
+  override def visitSelectStmt(ctx: SelectStmtContext): PSelectStmt = {
+    val clauses = ctx.commClause().asScala.toVector.map(visitCommClause)
+    // TODO: Do this in one pass
+    val send = clauses collect { case v: PSelectSend => v }
+    val rec = clauses collect { case v: PSelectRecv => v }
+    val arec = clauses collect { case v: PSelectAssRecv => v }
+    val srec = clauses collect { case v: PSelectShortRecv => v }
+    val dflt = clauses collect { case v: PSelectDflt => v }
+
+    PSelectStmt(send, rec, arec, srec, dflt).at(ctx)
+  }
+
+  /**
+    * {@inheritDoc  }
+    *
+    * <p>The default implementation returns the result of calling
+    * {@link #   visitChildren} on {@code ctx}.</p>
+    */
+  // TODO : Refactor
+  override def visitCommClause(ctx: CommClauseContext): PSelectClause = {
```

Comment:
Again, it seems here you wanted to refactor something.

Reply:
I assume, as with the previous point, we leave this for now, which is fine. I just wanted to bring it to your attention, in case you forgot it.